Stock items job

BenMorganIO opened the pull request:
In the previous PR, some bright guy (me) decided to merge master into the stock items job PR. That resulted in it being closed by Github after a revert and a hard reset. This PR takes off where #5888 left off.

This PR includes quite a bit of refactoring in the specs while searching for reasons behind some Fs.

BenMorganIO committed:
move stock items creation to a job

BenMorganIO committed:
Merge branch 'rails-4-2' into stock-items-job

BenMorganIO committed:
specs for Spree::CreateStockItemsJob

BenMorganIO committed:
use some spec shorthand to simplify inventory unit spec

BenMorganIO committed:
move jobs up, alphabetical ordering

BenMorganIO committed:
simplify code on inv unit spec more

BenMorganIO committed:
remove duplicate test

This is defined just a few lines higher. Removing.

BenMorganIO committed:
more inv unit spec refactors

BenMorganIO committed:
remove additional additional_tax_total spec

BenMorganIO committed:
move after commit below after save

An after commit happens after an after save. Visually represent this.

BenMorganIO committed:
add after commit callback runs to fix 2 specs

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://benmorganio-stock-items-job.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://benmorganio-stock-items-job.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

BenMorganIO commented on core/spec/models/spree/customer_return_spec.rb:
```diff
@@ -113,36 +108,31 @@
-        expect do
-          create(:customer_return_without_return_items, return_items: [return_item], stock_location_id: inventory_unit.shipment.stock_location_id)
-        end.to change { inventory_unit.find_stock_item.count_on_hand }.by(1)
+        create(:customer_return_without_return_items, return_items: [return_item], stock_location_id: inventory_unit.shipment.stock_location_id)
```
That's better than the 146...

BenMorganIO committed:
simplify #find_stock_item

BenMorganIO committed:
remove self and question mark

CircleCI seems to be crashing, not sure why.

BenMorganIO committed:
use a try for propagate_all_variants

In the migrations, we attempt to seed the data. This is done (I think)
before the `propagate_all_variants` column has been made (which comes
from a later migration).

To combat this, I’m adding a try, but ideally we should move the seed
data out from the migration and probably into the seeds.

JDutil commented on core/spec/spec_helper.rb:
```diff
@@ -4,10 +4,10 @@
     add_group 'Models', 'app/models'
     add_group 'Views', 'app/views'
-    add_group 'Jobs', 'app/jobs'
     add_group 'Libraries', 'lib'
```
Libraries can be alphabetized too.

JDutil commented on core/app/models/spree/stock_location.rb:
```diff
@@ -12,8 +12,8 @@ class StockLocation < Spree::Base
 
-    after_create :create_stock_items, :if => "self.propagate_all_variants?"
     after_save :ensure_one_default
+    after_commit :create_stock_items, on: :create, if: "self.try :propagate_all_variants?"
```
I'm not sure I like making this after_commit the stock location should properly be created so you shouldn't need the try.  I feel like this let's things possibly fail silently.

BenMorganIO replied:
I couldn't figure out how to get the tests passing without it...

I'm thinking of moving the seed data inside of the migrations out into the seeds.

JDutil replied:
Could have the seeds destroy the migrated data and seed fresh records if seeds are run.

BenMorganIO committed:
ruby 1.9 syntax

BenMorganIO committed:
fix broken spec from after commit

BenMorganIO committed:
move libraries to alphabetical position

JDutil commented:
Could you please resubmit against master?  Didn't notice this before merging 4.2 branch.
